Make PGO code state not sensitive to file path by hashing file content when the file is available.

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #152628

In some internal frameworks, on second attempts the actual code is copied to a different path than previous attempts. 
but its still the same. PGO will not work on those cased due to the following, sate entries before this PR used to be identified by (filepath, function name, line number).

after this PR they are identified by (hash(filepath) , function name, line number). This way PGO will work for those jobs on future attempts and re-compilations of static versions will be avoided.

Sometimes we do not have access to the source code, (file does not exists)
This seems to happen mostly when we re-trace a compiled function but generally it can happen . 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

laithsakka committed:
draft

[ghstack-poisoned]

laithsakka committed:
Update on "Make PGO code state non sensitive to file path by hashing file content. "


In some internal frameworks, on second attempts the actual code is copied to a different path than previous attempts. 
but its still the same. PGO will not work on those cased due to the following, sate entries before this PR used to be identified by (filepath, function name, line number).

after this PR they are identified by (hash(filepath) , function name, line number). This way PGO will work for those jobs on future attempts and re-compilations of static versions will be avoided.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka committed:
Update on "Make PGO code state non sensitive to file path by hashing file content. "


In some internal frameworks, on second attempts the actual code is copied to a different path than previous attempts. 
but its still the same. PGO will not work on those cased due to the following, sate entries before this PR used to be identified by (filepath, function name, line number).

after this PR they are identified by (hash(filepath) , function name, line number). This way PGO will work for those jobs on future attempts and re-compilations of static versions will be avoided.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

oulgen commented on torch/_dynamo/pgo.py:
```diff
@@ -103,15 +104,46 @@ class ReservedWorkflowIdUserError(ValueError):
+        if self.filename not in CodeId._hash_cache:
+            with open(self.filename, "rb") as file:
+                content = file.read()
+                crc32_value = binascii.crc32(content)
```
use zlib.crc32 instead

laithsakka replied:
done , what would be the difference. perf?

oulgen commented on torch/_dynamo/pgo.py:
```diff
@@ -103,15 +104,46 @@ class ReservedWorkflowIdUserError(ValueError):
+            with open(self.filename, "rb") as file:
+                content = file.read()
+                crc32_value = binascii.crc32(content)
+                hash = format(crc32_value & 0xFFFFFFFF, "08x")
```
why do you need this? crc32 should already give you 4 byte value?

laithsakka replied:
for better visualization

oulgen commented on torch/_dynamo/pgo.py:
```diff
@@ -103,15 +104,46 @@ class ReservedWorkflowIdUserError(ValueError):
+    def __hash__(self) -> int:
+        return hash((self.file_hash, self.name, self.firstlineno))
+
+    def _hash_containing_file(self) -> str:
```
you have essentially re-implemented functools.cache here, can you use that instead where this is a free function an argument is file name?

laithsakka replied:
done

oulgen commented on torch/_dynamo/pgo.py:
```diff
@@ -103,15 +104,46 @@ class ReservedWorkflowIdUserError(ValueError):
-    filename: str
-    firstlineno: int
-    name: str
+    _hash_cache: dict[str, str] = {}
```
if you use functool.lru_cache as free function this can remain to be a drozen dataclass, and you get hash and str for free

laithsakka committed:
Update on "Make PGO code state non sensitive to file path by hashing file content. "


In some internal frameworks, on second attempts the actual code is copied to a different path than previous attempts. 
but its still the same. PGO will not work on those cased due to the following, sate entries before this PR used to be identified by (filepath, function name, line number).

after this PR they are identified by (hash(filepath) , function name, line number). This way PGO will work for those jobs on future attempts and re-compilations of static versions will be avoided.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

oulgen commented on torch/_dynamo/pgo.py:
```diff
@@ -103,15 +105,43 @@ class ReservedWorkflowIdUserError(ValueError):
+        return hash
+
+
 class CodeId:
```
so now you can go back to frozen dataclass, son you dont have to implement hash and str. You can still manually implement eq.

laithsakka replied:
I can add the frozen dataclass back but i still need str, and hash 

for str i still want it, i do not want this representation
 return f"hash({self.file_hash}){self.filename}:{self.firstlineno}:
 which will be different.
 
for hashing if two objects are equal they should have the same hash so i still need it .

oulgen replied:
I'm not following your current hash implementation is
```
return hash((self.file_hash, self.name, self.firstlineno))
```
which is equal to what default dataclass would do.

laithsakka replied:
I am excluding self.file_name

oulgen replied:
ah you dropped filename

laithsakka replied:
added a comment closer to explain

oulgen commented on torch/_dynamo/pgo.py:
```diff
@@ -322,8 +352,9 @@ def update_automatic_dynamic(
-    code_id = CodeId.make(tx.f_code)
+    code_id = CodeId(tx.f_code)
     frame_state = get_code_state()[code_id]
+    print
```
delete

oulgen commented on torch/_dynamo/pgo.py:
```diff
@@ -557,7 +588,7 @@ def get_remote_cache() -> Optional[RemoteCache[JsonDataTy]]:
 def render_code_state(cs: defaultdict[CodeId, CodeState]) -> str:
     return "\n".join(
-        f"{k.filename}:{k.firstlineno}:{k.name}:\n"
+        f"{str(k)}:\n"
```
you dont need to call str, it will be automatic

oulgen commented on torch/_dynamo/pgo.py:
```diff
@@ -103,15 +105,43 @@ class ReservedWorkflowIdUserError(ValueError):
-        return CodeId(code.co_filename, code.co_firstlineno, code.co_name)
+    # Ensure if two CodeIds are the same, then they have the same hash by execluding filename.
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, CodeId):
+            return False
```
do you need this isinstance check? i dont think eq would be called unless they are same instance already

laithsakka committed:
Update on "Make PGO code state non sensitive to file path by hashing file content. "


In some internal frameworks, on second attempts the actual code is copied to a different path than previous attempts. 
but its still the same. PGO will not work on those cased due to the following, sate entries before this PR used to be identified by (filepath, function name, line number).

after this PR they are identified by (hash(filepath) , function name, line number). This way PGO will work for those jobs on future attempts and re-compilations of static versions will be avoided.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka committed:
Update on "Make PGO code state non sensitive to file path by hashing file content. "


In some internal frameworks, on second attempts the actual code is copied to a different path than previous attempts. 
but its still the same. PGO will not work on those cased due to the following, sate entries before this PR used to be identified by (filepath, function name, line number).

after this PR they are identified by (hash(filepath) , function name, line number). This way PGO will work for those jobs on future attempts and re-compilations of static versions will be avoided.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

oulgen commented on torch/_dynamo/pgo.py:
```diff
@@ -103,15 +105,53 @@ class ReservedWorkflowIdUserError(ValueError):
+    # code state it will refer to the first seen file path.
+    file_hash: str
+
+    # execlude file name.
```
```suggestion
    # exclude file name.
```

oulgen commented on torch/_dynamo/pgo.py:
```diff
@@ -103,15 +105,53 @@ class ReservedWorkflowIdUserError(ValueError):
+            and self.name == other.name
+        )
+
+    # Ensure if two CodeIds are the same, then they have the same hash by execluding filename.
```
```suggestion
    # Ensure if two CodeIds are the same, then they have the same hash by excluding filename.
```

laithsakka committed:
Update on "Make PGO code state non sensitive to file path by hashing file content. "


In some internal frameworks, on second attempts the actual code is copied to a different path than previous attempts. 
but its still the same. PGO will not work on those cased due to the following, sate entries before this PR used to be identified by (filepath, function name, line number).

after this PR they are identified by (hash(filepath) , function name, line number). This way PGO will work for those jobs on future attempts and re-compilations of static versions will be avoided.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka committed:
Update on "Make PGO code state non sensitive to file path by hashing file content when the file is available. "


In some internal frameworks, on second attempts the actual code is copied to a different path than previous attempts. 
but its still the same. PGO will not work on those cased due to the following, sate entries before this PR used to be identified by (filepath, function name, line number).

after this PR they are identified by (hash(filepath) , function name, line number). This way PGO will work for those jobs on future attempts and re-compilations of static versions will be avoided.

Sometimes we do not have access to the source code, (file does not exisits)
This seems to happen mostly when we re-trace a compiled function. 
cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

oulgen commented on torch/_dynamo/pgo.py:
```diff
@@ -103,15 +105,57 @@ class ReservedWorkflowIdUserError(ValueError):
+@functools.cache
+def _hash_containing_file(filepath: str) -> str:
+    # if the file does not exists we consider filepath to be the hash.
+    if not os.path.exists(filepath):
+        return filepath
+
```
this is scary, why would the file not exist?

laithsakka replied:
i looked at several unit tests that were failing due to this and for those we used to trace an exported program or pre compiled program. 

basically here if the file is not found the existing behaviour will be used. so this should not introduce new behaviour.

laithsakka replied:
the pattern is sometimes we graph break and we retrace a function that is generated on the fly. 
see for example
https://github.com/pytorch/pytorch/actions/runs/14782545802/job/41505517250

laithsakka replied:
this another example that I found in omniFM 
```
<string>:2:__init__:
  L['losses']: tensor size=[30] stride=[1]
  L['predictions']: tensor size=[30, 1152, 1] stride=[S(1), S(2), 1]
  L['labels']: tensor size=[30, 1152] stride=[S(1), 1]
  L['weights']: tensor size=[30, 1152] stride=[S(1), 1]
  ```
  you can see the path here is <string>
  
  I checked the IG one there is no such examples, those are rare i think its ok to tream them like this, 
  the way we can do is hash the co_code for those... but i think its fine

laithsakka commented:
@pytorchbot merge
